traitsui/qt4/tree_editor.py

```diff
@@ -1317,13 +1314,18 @@ def _on_nid_changed(self, nid, col):
             else:
                 self._set_label(nid, col)
 
-    def _menu_new_node(self, class_name, prompt=False):
+    def _menu_new_node(self, factory, prompt=False):
         """ Adds a new object to the current node.
         """
         node, object, nid = self._data
         self._data = None
-        new_node, new_class = self._node_for_class_name(class_name)
-        new_object = new_class()
+        try:
+            new_object = factory()
+        except:
+            from traitsui.api import raise_to_debug
+            raise_to_debug()
+            return
```

Comment:
This `try: ... except: ...` was necessary since without it the new code changes would cause an exception to be raised if one tried to create a new node for an object with required traits.  Previously the hacky eval swallowed this here:
https://github.com/enthought/traitsui/blob/a4be4b539be4dfae295525f86ed18f04ad67bc71/traitsui/qt4/tree_editor.py#L1203-L1219

Once we complete the second portion of this fix, namely "You then sit down and think about how you need to change the tree node API so that it can better specify non-class factories (ie. an improvement or alternative to get_add and get_name). You basically need some way to specify the name to display in the menu, and the factory function to call for each type of thing you want to add. There may even be something like this already in the API. The default implementation of this should just do what the current code does, but it should allow developers to have their own tree node classes supply alternatives." 
we will want to remove this.  At that point a programmer will have all the tools needed to specify how to create a new node that has required arguments, and if they do so in a way that we still can not create a node we should complain and raise an exception.  However now, with the machinery not yet in place, we don't want this to fail.  This preserves the current behavior.